### PR TITLE
Provide correct instructions how to upgrade Nix without nixpkgs.

### PR DESCRIPTION
**Pull request by domenkozar**
See https://github.com/NixOS/nixpkgs/issues/15344 https://github.com/NixOS/nix/pull/1021

### CODE REVIEW

**Comment by dezgeg** [comment]
This seems to duplicate these paths in `nixos-install.sh`: https://github.com/NixOS/nixpkgs/blob/582313bafef4c81cb6df2dcf2ece4757eb5c8082/nixos/modules/installer/tools/nixos-rebuild.sh#L257

Also would be better to figure out a way to automate that, like `nix-env -i $(curl -L -H "Accept: application/json" 'http://hydra.nixos.org/job/nixpkgs/trunk/nixStable.x86_64-linux/latest' | some magic here)` (possibly by enhancing Hydra to also give some easily shell-parseable output)

**Comment by copumpkin** [comment]
Or how about this: Hydra publishes a tarball to a canonical URL with a single nix file in it containing a string with the instructions in it.

Then this default.nix calls `fetchTarball` on that path (no hash needed) if the version is too low (or provides a link to it if `fetchTarball` is missing), evaluates the contents, and then prints out the message as an error.

**Comment by domenkozar** [comment]
That requires Hydra to be online while currently you just need cache.nixos.org to be there and I've never seen it down.

**Comment by copumpkin** [comment]
@domenkozar does it? I'm just saying Hydra uploads a file to a canonical URL (it could be anywhere for all I care) after a successful build, just like it updates channels on nixos.org.

**Comment by domenkozar** [comment]
@copumpkin for example https://nixos.org/releases/nix/latest/? I'm +1 doing this better, but let that be second iteration :)

**Comment by copumpkin** [comment]
Yeah, that. Definitely fine to do this for now, just don't want to have long-term instructions that will rot and mislead newbies 😄

**Comment by domenkozar** [comment]
Removing dependency on wiki is exactly one step to that goal :) Fully agreed.

**Comment by domenkozar** [comment]
One thing I've noticed, it's possible to install Nix for the wrong platform and completely break it. cc @edolstra

**Comment by copumpkin** [comment]
Now that you point that out, @domenkozar, it seems like with the thing I'm describing, the tarball could contain one attribute per `system`, which would then serve up the correct instructions based on the actual Nix that's evaluating it.

Again, not something we need to do now, but it'll cut out one potential human screwup eventually 🐱

**Comment by domenkozar** [comment]
@copumpkin good idea, I've now refactored that to print

```
$ nix-build 
error: evaluation aborted with the following error message: ‘This version of Nixpkgs requires Nix >= 1.20, pleas
e upgrade:

On your platform x86_64-linux:

  $ nix-env -i /nix/store/mln2lswvkyx6x7j6pcx80cyf06fsc12m-nix-1.11.2


’
```

File: default.nix [hunk 1]
```diff
@@ -1,8 +1,31 @@
...
+    i686-linux = "/nix/store/fpa7k4ra0jcy1jq3cig5738dmmqsqyjc-nix-1.11.2";
+    x86_64-darwin = "/nix/store/mcfhmajqw93xs4bgk55822gag11qs7yk-nix-1.11.2";
+  };
+  installMsg = if (hasAttr currentSystem platformPath) then ''
+    On your platform ${currentSystem}:
+
+      $ nix-env -i ${getAttr currentSystem platformPath}
+  '' else ''
+    Your platform ${currentSystem} doesn't have a binary cache. Please
+    reinstal Nix following instructions on http://nixos.org/nix/download.html
```
Comment (matthewbauer) [inline default.nix]: *reinstall

**Comment by domenkozar** [comment]
If there are no objections from @edolstra I'll merge this tomorrow.

**Comment by edolstra** [comment]
I don't think this is a good idea, because how to upgrade Nix is system-dependent. E.g. on NixOS, installing Nix via nix-env will cause interference with the version installed in the system profile. Likewise when Nix is installed via RPM/apt.

Also, this duplicates the store paths in `nixos-rebuild`, which means yet another manual step to perform when doing a Nix release.

Maybe we should just remove the wiki link?

**Comment by domenkozar** [comment]
@edolstra doesn't `nixos-rebuild` handle Nix upgrade automatically so user shouldn't never get this error?

I think it's common for folks not to upgrade Nix and we have to have a way for them to do so when nixpkgs requirement is bumped.

Maybe Nix should have some autoupgrading logic in there? Or at least warn that newer version is around.

For example pip (python installer) checks for newer versions on invocations and caches that for a while:

```
You are using pip version 8.0.2, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

I think this is "good enough" until we find a better way.

It's very important for Nix installation/usage not to introduce blockers since that gives people a sense of unstable package manager, leaving them in the paradox.

**Comment by copumpkin** [comment]
@edolstra I've run into several people who have encountered the "nix is too old" message, and have even found the current wiki page confusing or misleading. Making that experience worse for them seems like the wrong direction.

**Comment by domenkozar** [comment]
A tiny bit out of context, but here is @edolstra feedback:

```
domenkozar: well, my objections still stand: 1) this adds yet another place where we hard-code Nix store paths; 2) the instructions would be misleading on NixOS or on any system where Nix is installed via a non-Nix mechanism (e.g. RPM) or in a multi-user installation
```

I'll fix 1) and look into 2).

**Comment by vcunat** [comment]
@domenkozar (I'm very late): `nixos-rebuild` has these fallbacks to use when it's unable to get (newer) nix the usual way: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/tools.nix#L41 ... but IMO we could well factor these out and use them on both places (and e.g. add a Darwin fallback).

**Comment by vcunat** [comment]
From the manual it seems that the bash-curl method is far the easiest way to get nix onto a Linux or Darwin machine: http://nixos.org/nix/manual/#ch-installing-binary I'd think in that case it's enough to perform upgrades of nix itself by `nix-env` and this PR would seem a good suggestion (perhaps after extending by a note that the way might not be suitable in all cases).

**Comment by vi** [comment]
Feedback requested on IRC:

Situation:

*  ‘nix-env --version’ -> `nix-env (Nix) 1.10`
* `nix-instantiate --eval '<nixpkgs>' -A lib.nixpkgsVersion` -> `"17.03pre94828.ecbb932"`
* `uname -om` -> `x86_64 GNU/Linux`
* Link on the [wiki](https://nixos.org/wiki/How_to_update_when_Nix_is_too_old_to_evaluate_Nixpkgs) is 404 not found

Working solution provided [on IRC](https://botbot.me/freenode/nixos/msg/75990927/):

`nix-env -i /nix/store/mln2lswvkyx6x7j6pcx80cyf06fsc12m-nix-1.11.2`

**Comment by domenkozar** [comment]
For the long-term solution see https://github.com/NixOS/nix/issues/1113, I'd still merge something like this for short-term.

**Comment by dezgeg** [comment]
I don't think the current proposal will work on NixOS, because you can get into problems where the nix-daemon is running a too old version and just using a newer nix evaluator doesn't help. How about something more along these lines:

````
This version of Nixpkgs requires Nix >= ${requiredVersion}, please upgrade:

   - If you are running NixOS, use `nixos-rebuild' to upgrade your system.

   - If you installed Nix using the install script (https://nixos.org/nix/install),
     it is safe to upgrade by running it again:

        curl https://nixos.org/nix/install | sh
````

**Comment by dezgeg** [comment]
This one doesn't seem to be progressing. I'm closing this and implemented my suggestion above in https://github.com/NixOS/nixpkgs/pull/24017.

**Pull request closed without merging**